src/compiler/crystal/codegen/primitives.cr

```diff
@@ -173,6 +174,120 @@ class Crystal::CodeGenVisitor
+               when :i32
+                 binary_overflow_fun "llvm.sadd.with.overflow.i32", llvm_context.int32
+               when :i64
+                 binary_overflow_fun "llvm.sadd.with.overflow.i64", llvm_context.int64
+               when :i128
+                 binary_overflow_fun "llvm.sadd.with.overflow.i128", llvm_context.int128
+               when :u8
+                 binary_overflow_fun "llvm.uadd.with.overflow.i8", llvm_context.int8
+               when :u16
+                 binary_overflow_fun "llvm.uadd.with.overflow.i16", llvm_context.int16
+               when :u32
+                 binary_overflow_fun "llvm.uadd.with.overflow.i32", llvm_context.int32
+               when :u64
+                 binary_overflow_fun "llvm.uadd.with.overflow.i64", llvm_context.int64
+               when :u128
+                 binary_overflow_fun "llvm.uadd.with.overflow.i128", llvm_context.int128
+               else
+                 raise "unreachable"
+               end
+
+    codegen_binary_overflow_check(llvm_fun, t, t1, t2, p1, p2)
+  end
+
+  # Generates a call to llvm_fun(p1, p2).
+  # t1, t2 are the original types of p1, p2.
+  # t is the super type of t1 and t2 where the operation is performed.
+  # llvm_fun returns {res, o_bit} where the o_bit signals overflow.
+  # The generated code also performs a range check and truncation of res
+  # in order to fit in the original type t1 if needed.
+  #
+  # ```
+  # %res_with_overflow = call {T, i1} <llvm_fun>(T %p1, T %p2)
+  # %res = extractvalue {T, i1} %res, 0
+  # %o_bit = extractvalue {T, i1} %res, 1
+  # ;; if T != T1
+  # %out_of_range = %res < T1::MIN || %res > T1::MAX ;; compare T1.range and %res
+  # br i1 or(%o_bit, %out_of_range), label %overflow, label %normal
+  # ;; else
+  # br i1 %o_bit, label %overflow, label %normal
+  # ;; end
+  #
+  # overflow:
+  # ;; codegen: raise OverflowError.new with caller's location
+  #
+  # normal:
+  # ;; if T != T1
+  # ;;   %res' is returned
+  # %res' = trunc T %res to T1
+  # ;; else
+  # ;;   %res is returned
+  # ;; end
+  # ```
+  private def codegen_binary_overflow_check(llvm_fun, t : IntegerType, t1, t2, p1, p2)
+    res_with_overflow = builder.call(llvm_fun, [p1, p2])
+
+    res = extract_value res_with_overflow, 0
+    o_bit = extract_value res_with_overflow, 1
+
+    if t != t1
+      overflow = or(o_bit, codegen_out_of_range(t1, t, res))
+    else
+      overflow = o_bit
+    end
+
+    codegen_raise_overflow_cond overflow
+    codegen_trunc_binary_op_result(t1, t2, res)
+  end
+
+  private def codegen_out_of_range(target_type : IntegerType, arg_type : IntegerType, arg)
+    min_value, max_value = target_type.range
+    # arg < min_value || arg > max_value
+    or(
+      codegen_binary_op_lt(arg_type, target_type, arg, int(min_value, target_type)),
+      codegen_binary_op_gt(arg_type, target_type, arg, int(max_value, target_type))
+    )
+  end
+
+  private def codegen_raise_overflow
```

Comment:
Since the generated code here is always the same and very common, I think it would be best to make it at least a single basic block for each function. Ideally it'd be a single `__crystal_overflow` function, with a single basic block for each function which is a call to that function. From my testing, passing `-D preview_overflow` seems to add about 50% to the build time, which isn't great.

Reply:
You mean to have a `__crystal_overflow` well-known function that will raise the exception? What I don't like about that is that the location of the exception will not be the location of the operator. It will be along the trace.

If the slow down comes from using the `MainVisitor` maybe there is a another way to emit that fixed set of code.

Or maybe we can reuse the main visitor. There were only a pair of calls to `@program.visit_main` in the code (one with a manually created main visitor and other with a custom visitor.

Reply:
I think that if a function/method doesn't have a location it's skipped from the trace. So you could try defining that function in the compiler (so it doesn't get a location) and then use it.

Reply:
@bcardiff finding the exact location of the overflowing operation already depends on the backtrace line numbers surely? So nothing changed.

Reply:
done at "Add __crystal_raise_overflow as well known function". I think we can leave that commit not squashed.

The times of the codegen went from:
* 11.83s (master) 
* 37.14s (without __crystal_raise_overflow)
* 12.66s (with __crystal_raise_overflow)

Reply:
Nice! And thank you @RX14 for the great idea!